src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const { type, keysToRemove, extraData } of actions) {
+    switch (type) {
+      case 'addProperties':
+        addProps(currentState, extraData);
+        break;
+      case 'removeProperties':
+        removeProps(currentState, keysToRemove);
```

Comment:
The functions `addProps` and `removeProps` are mutating the `currentState` object directly. This violates the immutability requirement from the checklist. Instead of modifying `currentState`, you should create a new object and return it from these functions.